gate-web/src/main/groovy/com/netflix/spinnaker/gate/security/oauth2/SpinnakerUserInfoTokenServices.groovy

```diff
@@ -68,20 +69,26 @@ class SpinnakerUserInfoTokenServices implements ResourceServerTokenServices {
   @Autowired
   Front50Service front50Service
 
+  @Autowired(required = false)
+  SpinnakerProviderTokenServices providerTokenServices
+
   @Override
   OAuth2Authentication loadAuthentication(String accessToken) throws AuthenticationException, InvalidTokenException {
     OAuth2Authentication oAuth2Authentication = userInfoTokenServices.loadAuthentication(accessToken)
 
     Map details = oAuth2Authentication.userAuthentication.details as Map
 
     if (log.isDebugEnabled()) {
-      log.debug("UserInfo details: ", entries(details))
+      log.debug("UserInfo details: " + entries(details))
     }
 
     def isServiceAccount = isServiceAccount(details)
     if (!hasAllUserInfoRequirements(details) && !isServiceAccount) {
       throw new BadCredentialsException("User's info does not have all required fields.")
     }
+    if (providerTokenServices != null && !providerTokenServices.hasAllProviderRequirements(accessToken, details)) {
```

Comment:
The original intent of the `hasAllUserInfoRequirements` checks above was to do something similar to what you're working toward - namely the ability to do some coarse-grain authorization checks. 

In the userInfoRequirement case, we're just checking that the user's profile contains a predefined string (or matches a regex). It seems like generalizing this capability is now needed for more complex use cases, such as the ability to hit a GitHub API to check org membership. 

Can you file an issue in spinnaker/spinnaker linking to this PR and suggesting that we consolidate these mechanisms into a generic way to check coarse grain authz at login time? We can then merge this and maybe @stewchen can tackle it.